docs/rfc/rfc-19-random-beacon-v2.adoc

```diff
@@ -0,0 +1,169 @@
+for staking providers and calculating the provider’s fee was complicated given
+the non-reimbursable cost of ticket submission. Surprisingly, it is much easier
+for staking providers to calculate their fee and costs in tBTC v1, having a
+clean rule that no operations are reimbursed and the only reward they receive
+is KEEP plus TBTC fees.
+
+=== Aggressive timeouts
+
+Relay entry timeouts were pretty aggressive. Set to 64x6 = 384 blocks, they were
+giving about 1.5 hours to submit a relay entry before the minimum stake from each
+group member was slashed. This timeout proved to be not enough to notify,
+investigate, and fix a problem. Three groups were slashed in a row during the
+November 11th incident that begun around 08:00 UTC, before any US-based
+operators noticed the problem.
+
+
+== Proposal
+
+[[fees-and-rewards]]
+=== Fees and rewards
+
+In tBTC v1, operators have to execute a lot of on-chain transactions per
+deposit. Even though they are not reimbursed for on-chain actions and they
+receive KEEP rewards and TBTC fees, this scheme proved to be successful and no
+operators complained about it.
+
+For v2 of the random beacon, all rewards and reimbursements should be done in T.
+This scheme is compatible with tBTC v2 design where all rewards and
+reimbursements are also done only in T.
+
+Relay requester should provide a fee in T. The value of the fee is a governable
+parameter. The fee is split into two parts. One part goes to a group creation
+pool and another part is used as a reimbursement for the relay entry submitter.
+What percent of the fee goes to the group creation pool is a governable
+parameter.
+
+Given that we want to incentivize the submission of a relay entry as soon as possible
+and that the price relation between gas price and T reimbursement for relay
+entry submission is unknown at the moment the fee is set by the governance, we
+should calculate a delay factor and delay penalty as in v1 of the random beacon.
+The delay penalty should be subtracted from reimbursement and sent to the
+sortition-pool-wide reward pool. The remaining part is sent to the relay entry
+submitter. Over the course of 64x6 blocks, reimbursement goes to zero. On-chain we
+are not going to enforce any order of submitters but off-chain clients should
+agree to the order where the member with index 1 has 6 blocks to submit, after 6
+blocks member with index 2 tries to submit, after 12 blocks member with index 3
+tries to submit and so forth.
+
+Anyone should be able to submit a transaction creating a new group and this
+action should be detached from relay entry submission. The submitter of this
+transaction receives a part of the group creation pool and the remaining part
+should be reserved for the DKG result submitter as reimbursement for a future
+transaction. What percentage of the pool goes where is a governable parameter
+and should be determined based on the gas cost of the two transactions.
```

Comment:
This is going to require some work to figure out with the new static sortition pool design. Ideally we want to lock the sortition pool before the entry for the new group creation can be known, so the exact sequence has some interactions that need sorting out.

Reply:
One option would be to have a governable number determining the frequency of creating new groups per the number of entries generated. With the new static sortition pool design, the transaction requesting to create a new group is no longer expensive so we can not use it to guess the price of T and gas and assume DKG result submission will be net positive.

Let's say that a new group would be created every 20 entries. When relay request n*20+1 is submitted to the chain, it marks that the new group creation is in progress and locks the sortition pool. Now, stakers execute sortition view function to determine who is selected and start the DKG. Once the result is published, the call to sortition pool is executed saying it to unblock the state once 100(?) additional blocks passes.

What do you think?

Reply:
> When relay request n*20+1 is submitted to the chain, it marks that the new group creation is in progress and locks the sortition pool.

If we lock the pool when someone _requests_ the group-creating entry, it would work fine. It would slightly increase the cost of requesting entries sometimes, but if we do the sortition pool correctly it should only be 6-7k gas + change.

However, we need to make sure that relay requests aren't submitted too often; if the pool is locked for an average of 120 blocks every DKG, if entries are requested every 6th block on average the pool would practically always be locked. We could enforce a minimum number of blocks or seconds between group creations to alleviate this risk faster than governance can respond.

Reply:
Good point. So let me try to summarize it:
- New group creation is triggered with a fixed (yet governable) frequency of entries, `dkg_freq`.
- When someone requests for a new entry and the counter says we should trigger DKG with the new entry  (we are at `dkg_freq - 1`), we lock the sortition pool.
- Sortition happens optimistically: the pool is locked, members are determined via a view function call, there is some time for complaints before the group is considered active, sortition pool is unlocked after the complaints period times out.
- Relay request transaction locking the pool, triggers new DKG.
- No one is getting reimbursed for triggering DKG.

Does it make sense?

In this model, we will have to figure out how to reward DKG result submitter - this model does not require a third party to submit expensive transaction so we can not estimate current market conditions based on it. 

Given that this action is not happening as often as relay entry submission but is much more expensive, maybe it is enough if we set a fixed, governable reward for submitting an entry and enforce order on-chain? This reward does not always need to cover the costs but might be used just to alleviate the pain. What do you think?

Reply:
Given that triggering DKG is just a simple check + state change (locking the pool) I agree that it can be wrapped into the entry request transaction.

For DKG result submission we could allocate a reward (possibly a fraction of the maintenance reward pool), but have the submitter reward grow over time from zero to the maximum so that higher gas costs would lead to DKG results being submitted slower and vice versa. Since the submission has to verify `threshold + 1` ECDSA signatures, plus create a new wallet and record the members, it's still relatively expensive. (We might be able to reduce the cost by encoding the operator IDs as `immutable` and only having a mutable bitstring denoting their status, but it would be more complex than densely packed storage slots.)

After the challenge period is over, one more transaction is needed for confirming the group and unlocking the sortition pool and releasing the submitter reward (we don't want to pay out rewards for fraudulent submissions). This again would be fairly cheap, and because it releases the reward the submitter would have an incentive to call it anyway. The group could be mildly griefed by the submitter not confirming it, but any other member (or indeed any address whatsoever) could do it just as well. We could also wrap the confirmation into the next entry request or submission after the challenge period is over, to make it happen automatically, but I think requiring a separate transaction is simpler and doesn't rely on entry frequency.

Reply:
Increasing the reward seems interesting but I am worried that the additional complexity may not be worth it. It is similar to relay entry submission - except that the time it takes to submit DKG result might not be that important - it may happen that even 100% of the reward does not cover the cost, and we do not want operators to race for submission and have 63 transactions reverted when the price reaches profitability point plus some bonus. 

What I think we could do: 
- have a fixed, governable reward for DKG result submission
- enforce the order on-chain, member 1 is eligible first; then 1 and 2; then 1, 2, 3; then 1, 2, 3, 4;...
- member who was supposed to submit the result but did not is removed from the sortition pool

The rest stays as you say: two transactions, first locking the pool, and second approving the result, sending the reward to submitter and unlocking the pool. There will be some timeout for DKG result submission and in case it was not submitted on time, anyone can unlock the pool and be rewarded for it. Locking the pool will be wrapped in the entry request transaction.